### PR TITLE
[HttpKernel] Fixed RequestDataCollector handling of null header values.

### PR DESCRIPTION
**Pull request by gmoreira**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

This PR references this [discussion](https://github.com/symfony/http-kernel/commit/1fee78443cfa1f147d8d0471c8ac109e9cf24854#commitcomment-20028839).

### CODE REVIEW

**Comment by ro0NL** [comment]
`$response->headers->set('X-Foo-Bar', null);` is invalid :confused: The API of `$values` is `string|string[]` (well.. `string|array` actually).

But it doesnt expect `null` at all. Hence `remove()`.

**Comment by gmoreira** [comment]
@ro0NL You are correct, I should have noticed this. However, this brings up an interesting point, should the HeaderBag set() method protect against this being possible?

**Comment by ro0NL** [comment]
No, this is contract by design. We could however clarify the PHPdoc, ie. `string|string[]`, and perhaps default an empty array to `array('')` for the sake of it.

Then there are also a few `array_key_exists` checks which shouldnt be needed imo.

**Comment by alcalyn** [comment]
Personnaly I don't use `$headersBag->set('X-Foo-Bar', null);`, but `$headersBag->set('X-Foo-Bar', $myValue);`, and unfortunately `$myValue` was sometimes `null`. It don't affect the behaviour until this bug.

I think that if we set null, it should just not appears in headers, and let an empty array in headers, like before `3.2`. So I agree this PR.

**Comment by ro0NL** [comment]
Imo setting `null` equals setting an empty string; ie. a header value is a string.

> and unfortunately $myValue was sometimes null

That doesnt mean you can just violate `string|string[]`.

**Comment by gmoreira** [comment]
@nicolas-grekas would be interested in your opinion here.

File: src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php [hunk 1]
```diff
@@ -125,7 +125,7 @@ public function collect(Request $request, Response $response, \Exception $except
                 continue;
             }
             if ('request_headers' === $key || 'response_headers' === $key) {
-                $value = array_map(function ($v) { return isset($v[1]) ? $v : $v[0]; }, $value);
+                $value = array_map(function ($v) { return isset($v[1]) ? $v : (isset($v[0]) ? $v[0] : null); }, $value);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php]: To be even more defensive, I'd suggest: `isset($v[0]) && !isset($v[1]) ? $v[0] : $v`

Comment (gmoreira) [inline src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php]: Sure. With regards to the unit test however, would you omit this change to the unit test in order to not violate the current interface contract of the HeaderBag::set() method.

Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php]: If it's green, let's do with for now.

**Comment by nicolas-grekas** [comment]
:+1:

**Comment by fabpot** [comment]
Is it really something for master only?

**Comment by nicolas-grekas** [comment]
3.2!

**Comment by fabpot** [comment]
Thank you @gmoreira.

**Pull request closed without merging**